### ISSUE TITLE
Performing Actions on recently created objects on a reference does not trigger `onSnapshot` for the store.

### ISSUE THREAD

**Issue by NickMele** [mobxjs/mobx-state-tree#993]
I have:

* [x] I think something is not working as it should.
  * [x] I've checked documentation and searched for existing issues
  * [x] I've made sure your project is based on the latest MST version
  * [x] Fork [this](https://codesandbox.io/s/jlnmjqr7xv) code sandbox or another minimal reproduction. 
  * [x] Describe expected behavior
  * [x] Describe observed behavior
   
I am running into an issue with performing an action on a reference not triggering the "onSnapshot" call in our store.

I am able to reproduce this in a simple store example: https://codesandbox.io/s/rrk8wojnp4

**Observed behavior:**
When loading the initial store, I want to perform an action on one selected todo. this selected todo is stored as a reference to a todo in the todos array. When I perform the `toggle` action using the reference you can see that there is no snapshot logged. Just a second later if I perform that same action directly on the item instead of via the reference, that action gets logged and can be stored.

**Expected behavior:**
Performing an action on the referenced todo should trigger a snapshot, just the same as performing that action directly on the item.

I am still pretty new to this library so I might be misunderstanding how references are supposed to work or be used.

**Comment by xaviergonz** [comment]
I'd say that's a bug, since strictly speaking "something" inside the store changed

**Comment by xaviergonz** [comment]
makes me wonder also if actions are recorded as well when doing an action over a reference

**Comment by NickMele** [comment]
@xaviergonz In my local project I am seeing the actions recorded in reactotron, and I can also get the patch when using `onPatch`. The only piece that does not seem to be working in `onSnapshot`

**Comment by xaviergonz** [comment]
yep, just tested the same with https://codesandbox.io/s/qv6k9l89jj

**Comment by xaviergonz** [comment]
A funny thing, if the object is modified a first time using the direct child *before* changing it using the reference then it works ok
https://codesandbox.io/s/qv6k9l89jj

@k-g-a might be something related to initialSnapshot?

**Comment by xaviergonz** [comment]
summarizing:
create -> change referenced model (no snapshot) -> change using child model (snapshot)
create -> change using child model (snapshot) -> change referenced model (snapshot) -> change using child model (snapshot)

**Comment by xaviergonz** [comment]
unit test
```ts
test("#993", () => {
    const Todo = types
        .model("Todo", {
            id: types.identifier,
            finished: false
        })
        .actions(self => ({
            toggle() {
                self.finished = !self.finished
            }
        }))

    const TodoStore = types.model("TodoStore", {
        todos: types.array(Todo),
        selectedTodo: types.reference(Todo)
    })

    const store = TodoStore.create({
        todos: [
            {
                id: "11124091-11c1-4dda-b2ed-7dd6323491a5"
            },
            {
                id: "23424091-11c1-4dda-b2ed-7dd6323491a6"
            }
        ],
        selectedTodo: "11124091-11c1-4dda-b2ed-7dd6323491a5"
    })

    let calls = 0
    onSnapshot(store, () => {
        calls++
    })

    expect(calls).toBe(0)
    store.selectedTodo.toggle()
    expect(calls).toBe(1)
    store.todos[0].toggle()
    expect(calls).toBe(2)
    store.selectedTodo.toggle()
    expect(calls).toBe(3)
})
```

**Comment by xaviergonz** [comment]
Apparently it has something to do with the array, I checked this other unit test and it didn't fail:
```ts
test("#993 without array, which is ok", () => {
    const Todo = types
        .model("Todo", {
            id: types.identifier,
            finished: false
        })
        .actions(self => ({
            toggle() {
                self.finished = !self.finished
            }
        }))

    const TodoStore = types.model("TodoStore", {
        todo: Todo,
        selectedTodo: types.reference(Todo)
    })

    const store = TodoStore.create({
        todo: {
            id: "11124091-11c1-4dda-b2ed-7dd6323491a5"
        },
        selectedTodo: "11124091-11c1-4dda-b2ed-7dd6323491a5"
    })

    let calls = 0
    onSnapshot(store, () => {
        calls++
    })

    expect(calls).toBe(0)
    store.selectedTodo.toggle()
    expect(calls).toBe(1)
    store.todo.toggle()
    expect(calls).toBe(2)
    store.selectedTodo.toggle()
    expect(calls).toBe(3)
})
```

**Comment by xaviergonz** [comment]
Ok, maps are affected as well, as well as objects with two levels deep, such as this one
```ts
test("#993", () => {
    const Todo = types
        .model("Todo", {
            id: types.identifier,
            finished: false
        })
        .actions(self => ({
            toggle() {
                self.finished = !self.finished
            }
        }))

    const TodoContainer = types.model({
        todo: Todo
    })

    const TodoStore = types.model("TodoStore", {
        todos: TodoContainer,
        selectedTodo: types.reference(Todo)
    })

    const store = TodoStore.create({
        todos: {
            todo: {
                id: "11124091-11c1-4dda-b2ed-7dd6323491a5"
            }
        },
        selectedTodo: "11124091-11c1-4dda-b2ed-7dd6323491a5"
    })

    let calls = 0
    onSnapshot(store, () => {
        calls++
    })

    expect(calls).toBe(0)
    store.selectedTodo.toggle()
    expect(calls).toBe(1)
    store.todos.todo.toggle()
    expect(calls).toBe(2)
    store.selectedTodo.toggle()
    expect(calls).toBe(3)
})
```
so apparently it happens when the reference and the referenced instance don't share a same parent?

**Comment by xaviergonz** [comment]
When I changed on that last unit test the last part to
```ts
    store.todos // this line makes it work
    expect(calls).toBe(0)
    store.selectedTodo.toggle()
    expect(calls).toBe(1)
    store.todos.todo.toggle()
    expect(calls).toBe(2)
    store.selectedTodo.toggle()
    expect(calls).toBe(3)
```
the test worked, so it seems to be about the lazy instantiation; store.selectedTodo does not seem to instantiate all the parents of its reference (store.todos in this case)

**Comment by xaviergonz** [comment]
Yep, `getParent(store.selectedTodo)` is undefined if `store.todos` is not called before, and an object if it is called before

**Comment by ovidius72** [comment]
When using `addMiddleware` to the root store its seems that it's not capturing event from the items attached as reference. Is it related ?

**Comment by xaviergonz** [comment]
@ovidius72 care to add a unit test or a codesandbox? I just created a potential fix, so I might add a test there to see if it fixes that

**Comment by xaviergonz** [comment]
It is not getting attached because you are creating a reference, but an actual Ref object with id "aaa" is not a _real_ child of the model
References are supposed to be used to point to children of the same model, this is:
```
ROOT
  - children of type C: [ child1, child2 ]
  - reference to type C: -> points to child1 (or 2)
```
in your case you have something like this
```
NO PARENT
- an instance of type C (1)

ROOT
  - reference to type C: -> points to (1), which doesn't get attached since it doesn't have a parent
```

**Comment by ovidius72** [comment]
@xaviergonz Ah, so i need to have a filed that stores "real" object and than another one for references ?

**Comment by xaviergonz** [comment]
@NickMele should be fixed in 3.3.0, feel free to reopen it if not!
